### DISCUSSION TITLE
Need help figuring out why my connections aren't closing with a fastapi dependency

### DISCUSSION THREAD

**Discussion by vamshiaruru-virgodesigns** [sqlalchemy/sqlalchemy#9246] [Usage Questions]
Hi all, firstly great job with 2.0 (though this question pertains to 1.4)! Secondly, I am not sure if this is the right place for it, since this might be an issue regarding my fastapi usage, but any help is appreciated. 

I have a dependency defined like this
```python
async def get_db() -> AsyncIterator[AsyncSession]:
    async with AsyncSessionLocal() as session:
        try:
            yield session
        except Exception as exc:
            logger.error(exc)
            logger.error(traceback.format_exc())
            await session.rollback()
            raise
        finally:
            await session.close()
```

where `AsyncSessionLocal` is 

```python

# Need this to connect via pgbouncer with transaction mode connection pooling
class CConnection(Connection):
    def _get_unique_id(self, prefix: str) -> str:
        return f"__asyncpg_{prefix}_{uuid4()}__"


connect_args = {
        "statement_cache_size": 0,
        "prepared_statement_cache_size": 0,
        "connection_class": CConnection,
}

engine = create_async_engine(
        settings.database_dsn,
        poolclass=NullPool,
        pool_pre_ping=True,
        connect_args=connect_args,
)

AsyncSessionLocal = sessionmaker(engine, class_=AsyncSession, expire_on_commit=False)  # type: ignore
```
I use the dependency like so:

```python
@router.post("/event-processor")
async def process_event(
    request: Request,
    event: Event,
    session: AsyncSession = Depends(get_db),
):
    # Do something with the session
```
I noticed however that the session / the underlying db connection isn't being closed properly. I see errors like this:
```
The garbage collector is trying to clean up connection <AdaptedConnection <api.db.session.CConnection object at 0x7eff882e22e0>>. This feature is unsupported on async dbapi, since no IO can be performed at this stage to reset the connection. Please close out all connections when they are no longer used, calling ``close()`` or using a context manager to manage their lifetime
```
And the active client connections on pgbouncer keeps increasing constantly to the point where there are no more client connections available on pg bouncer and I have to restart my containers to bring the active client connections down. If I don't use the dependency, and use the `async with AsynSessionLocal() as session` to get the session, it works properly. Does anyone know why this might be happening? Is it because of the `await session.close()`? 

Any help is appreciated, thanks.

**Comment by zzzeek** [comment]
not sure, can you try to isolate more specific parts of your program that cause the problem?  That is, first try connecting to a local database, and not pgbouncer.  Then try not not using connection_class.  Etc.    The pgbouncer part of this would not be related to the pooled connections not being returned, but perhaps something is going wrong elsewhere and that's a side effect.

the main thing to try is reproducing without pgbouncer actually in place. then you can create a reproducer and share here.

**Comment by CaselIT** [comment]
>  If I don't use the dependency, and use the async with AsynSessionLocal() as session to get the session, it works properly. Does anyone know why this might be happening? Is it because of the await session.close()?

Does using your `get_db` directly work here? You wrap it in an `asynccontextmanager` to make it usable directly:

```py
get_db_ctx = asynccontextmanager(get_db)

async def process_event(...):
  with get_db_ctx() as session:
    ...
```

does the following work?
